### PR TITLE
Fix native cache to support non-global IDs

### PR DESCRIPTION
**Pull request by ahmadbaraka**
Changes in this PR address issue #7 to fix the native code to match the JS and support non-global IDs caching

### CODE REVIEW

**Comment by radex** [comment]
Nice work @ahmadbaraka! 

There are some issues:
- See CI output: https://travis-ci.com/Nozbe/WatermelonDB/jobs/162152637 — Swift doesn't compile and there are Flow errors
- It would be good to have at least a basic test verifying that both adapter.query and adapter.find actually behave correctly. This is the file: https://github.com/Nozbe/WatermelonDB/blob/master/src/adapters/__tests__/commonTests.js

**Comment by ahmadbaraka** [comment]
@radex Thanks for quick feedback.
Some changes were done by mistake on the `dev` version in my app instead of being done in the source repo, so they were missed in the commit.

Currently waiting for CI to complete to verify the fixes.

----

> It would be good to have at least a basic test verifying that both adapter.query and adapter.find actually behave correctly. This is the file: https://github.com/Nozbe/WatermelonDB/blob/master/src/adapters/__tests__/commonTests.js

Regarding the tests, I see that file has common tests used for the adapters, but the changes here include implementation details of the SQLite adapter. I see the current tests also verify that `query` and `find` would return either `id` if the model is already cached, or `model._raw` if it's not cached.
Could you point out which test cases we can add here?

**Comment by radex** [comment]
> Regarding the tests, I see that file has common tests used for the adapters, but the changes here include implementation details of the SQLite adapter. I see the current tests also verify that query and find would return either id if the model is already cached, or model._raw if it's not cached.
> Could you point out which test cases we can add here?

Right, don't test the implementation details, test the behavior. I'd add a test for both query and find, checking that if I query/find "xxxx" in table A, which is already cached, it will return "xxxx", but if I do the same with table B (where it's not cached), it still returns the whole record.

PS. You only made the change in native implementation, not in LokiJS/web implementation. So you'll have to either also add the LokiJS implementation (should be easy), or, you can just wrap the test in is-native check (in which case, please mark in changelog that this is PARTIAL support for non-global IDs - native only)

**Comment by ahmadbaraka** [comment]
I've pushed additional changes to address your comment:
1. Added the LokiJS implementation for non-global IDs
1. Implemented tests for the new changes for `find` and `query` as suggested

File: native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt [hunk 1]
```diff
@@ -129,29 +129,29 @@ class DatabaseDriver(context: Context, dbName: String) {
...
-                        removedIds.add(it.id)
+                        removedIds.add(Pair(it.table, it.id))
                     }
                 }
             }
         }
-        newIds.forEach(this::markAsCached)
-        removedIds.forEach { cachedRecords.remove(it) }
+        newIds.forEach { markAsCached(it.first, it.second) }
+        removedIds.forEach { removeFromCache(it.first, it.second) }
```
Comment (rozPierog) [inline native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt]:
```suggestion
        removedIds.forEach { removeFromCache(table = it.first, id = it.second) }
```

File: native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt [hunk 2]
```diff
@@ -163,12 +163,19 @@ class DatabaseDriver(context: Context, dbName: String) {
...
     fun close() = database.close()
 
-    private fun markAsCached(id: RecordID) {
+    private fun markAsCached(table: TableName, id: RecordID) {
         log?.info("Mark as cached $id")
-        cachedRecords.add(id)
+        var cache = cachedRecords[table]
+        if (cache == null) {
+            cache = mutableListOf<RecordID>()
+            cachedRecords[table] = cache!!
```
Comment (rozPierog) [inline native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt]: `mutableListOf<RecordID>()` is creating non-nullable type of object so double bang is unnecessary

File: native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt [hunk 3]
```diff
@@ -163,12 +163,19 @@ class DatabaseDriver(context: Context, dbName: String) {
...
+        }
+        cache!!.add(id)
```
Comment (rozPierog) [inline native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt]: Same here

File: native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt [hunk 4]
```diff
@@ -163,12 +163,19 @@ class DatabaseDriver(context: Context, dbName: String) {
 
     fun close() = database.close()
 
-    private fun markAsCached(id: RecordID) {
+    private fun markAsCached(table: TableName, id: RecordID) {
         log?.info("Mark as cached $id")
-        cachedRecords.add(id)
+        var cache = cachedRecords[table]
+        if (cache == null) {
+            cache = mutableListOf<RecordID>()
```
Comment (rozPierog) [inline native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt]: Why do you state Type of objects in list if you are leaving it empty?

File: native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt [hunk 5]
```diff
@@ -129,29 +129,29 @@ class DatabaseDriver(context: Context, dbName: String) {
...
                         database.execute(Queries.destroyPermanently(it.table), arrayListOf(it.id))
-                        removedIds.add(it.id)
+                        removedIds.add(Pair(it.table, it.id))
                     }
                 }
             }
         }
-        newIds.forEach(this::markAsCached)
-        removedIds.forEach { cachedRecords.remove(it) }
+        newIds.forEach { markAsCached(it.first, it.second) }
```
Comment (rozPierog) [inline native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt]:
```suggestion
        newIds.forEach { markAsCached(table = it.first, id = it.second) }
```

File: native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt [hunk 6]
```diff
@@ -163,12 +163,19 @@ class DatabaseDriver(context: Context, dbName: String) {
...
+            cache = mutableListOf<RecordID>()
+            cachedRecords[table] = cache!!
+        }
+        cache!!.add(id)
     }
 
-    private fun isCached(id: RecordID): Boolean = cachedRecords.contains(id)
+    private fun isCached(table: TableName, id: RecordID): Boolean = cachedRecords[table]?.contains(id) ?: false
+
+    private fun removeFromCache(table: TableName, id: RecordID) { cachedRecords[table]?.remove(id) }
```
Comment (rozPierog) [inline native/android/src/main/java/com/nozbe/watermelondb/DatabaseDriver.kt]: Why not:
```suggestion
    private fun removeFromCache(table: TableName, id: RecordID) = cachedRecords[table]?.remove(id)
```

File: native/ios/WatermelonDB/DatabaseDriver.swift [hunk 7]
```diff
@@ -163,14 +163,23 @@ class DatabaseDriver {
...
 
-    private func markAsCached(_ id: RecordId) {
-        cachedRecords.insert(id)
+    private func markAsCached(_ table: Database.TableName, _ id: RecordId) {
+        var cachedSet = cachedRecords[table];
+        if cachedSet == nil {
+            cachedSet = []
+        }
+        cachedSet!.insert(id)
+        cachedRecords[table] = cachedSet
```
Comment (radex) [inline native/ios/WatermelonDB/DatabaseDriver.swift]: Let's do this without force unwrapping. How about:

```
var cachedSet = cachedRecords[table] ?? []
cachedSet.insert(id)
cachedRecords[table] = cachedSet
```

File: src/adapters/__tests__/commonTests.js [hunk 8]
```diff
@@ -112,6 +114,72 @@ export default () => [
...
+      adapter = adapter.testClone()
+
+      const s2 = makeMockTask({ id: 'id2', text1: 'baz', order: 2 })
+      await adapter.batch([['create', s2]])
+
+      // returns cached IDs after create
+      expectSortedEqual(await adapter.query(taskQuery()), [s1._raw, 'id2'])
+
+      // returns raw if not cached for a different table
+      expectSortedEqual(await adapter.query(projectQuery()), [p1._raw, p2._raw])
```
Comment (radex) [inline src/adapters/__tests__/commonTests.js]: can you check that the same call the second time will return ids? you're only testing caching after insert works, not that caching after query works

File: src/adapters/__tests__/commonTests.js [hunk 9]
```diff
@@ -112,6 +114,72 @@ export default () => [
...
+      adapter = adapter.testClone()
+
+      const s2 = makeMockTask({ id: 'id2', text1: 'baz', order: 2 })
+      await adapter.batch([['create', s2]])
+
+      // returns cached ID after create
+      expect(await adapter.find('tasks', 'id2')).toBe('id2')
+
+      // returns raw if not cached for a different table
+      expect(await adapter.find('projects', 'id2')).toEqual(p2._raw)
```
Comment (radex) [inline src/adapters/__tests__/commonTests.js]: can you check that the same call the second time will return ids? you're only testing caching after insert works, not that caching after find works

File: src/adapters/lokijs/worker/executor.js [hunk 10]
```diff
@@ -56,8 +56,36 @@ export default class LokiExecutor {
     await this._migrateIfNeeded()
   }
 
+  isCached(table: TableName<any>, id: RecordId): boolean {
+    const cachedSet = this.cachedRecords.get(table)
+    if (!cachedSet) {
+      return false
+    }
+
+    return cachedSet.has(id)
```
Comment (radex) [inline src/adapters/lokijs/worker/executor.js]: i think we can nicely inline to:

```
return cachedSet ? cachedSet.has(id) : false
```

File: src/adapters/lokijs/worker/executor.js [hunk 11]
```diff
@@ -56,8 +56,36 @@ export default class LokiExecutor {
...
+    cachedSet.add(id)
+  }
+
+  removeFromCache(table: TableName<any>, id: RecordId): void {
+    const cachedSet = this.cachedRecords.get(table)
+    if (!cachedSet) {
+      return
+    }
+
+    cachedSet.delete(id)
```
Comment (radex) [inline src/adapters/lokijs/worker/executor.js]: instead of early return, do the delete inside if

File: src/adapters/lokijs/worker/executor.js [hunk 12]
```diff
@@ -56,8 +56,36 @@ export default class LokiExecutor {
...
+  }
+
+  markAsCached(table: TableName<any>, id: RecordId): void {
+    let cachedSet = this.cachedRecords.get(table)
+    if (!cachedSet) {
+      cachedSet = new Set()
+      this.cachedRecords.set(table, cachedSet)
+    }
+
+    cachedSet.add(id)
```
Comment (radex) [inline src/adapters/lokijs/worker/executor.js]: let's do it without local mutability:

cachedSet = xxxx

if ( cachedSet) {
cachedSet.add(x)
} else {
cachedRecords.set(table, new Set(x))

**Comment by ahmadbaraka** [comment]
> @rozPierog take a look please ;) — could we avoid mutability in the Kotlin code similar to swift/js?

Fixed

**Comment by rozPierog** [comment]
@ahmadbaraka can you fix conflict in `CHANGELOG`? Thanks!

**Comment by ahmadbaraka** [comment]
@rozPierog I've resolved the conflict and done a rebase

**Comment by radex** [comment]
Cool! I'll merge as soon as the CI status turns green!

@ahmadbaraka Do y'all want some Watermelon stickers? If you do, send me an email and I'll ship you some!

**Comment by ahmadbaraka** [comment]
@radex I've sent you an email.

Is the failed build caused somehow by the changes in this PR?

**Comment by radex** [comment]
Merging. Thanks @ahmadbaraka — nice work!

**Comment by Kabangi** [comment]
When is this expected to be available on NPM?

**Comment by radex** [comment]
@Kabangi after New Year — I don't want to break something while I'm away from work. You can just use @ahmadbaraka's fork for now — you'll find the instructions to build production package in Contributing.md

**Pull request merged**